clients/src/main/java/org/apache/kafka/common/network/SslTransportLayer.java

```diff
@@ -420,6 +445,11 @@ private SSLEngineResult handshakeUnwrap(boolean doRead) throws IOException {
             log.trace("SSLHandshake handshakeUnwrap: handshakeStatus {} status {}", handshakeStatus, result.getStatus());
         } while (netReadBuffer.position() != 0 && cont);
 
+        // Throw EOF exception for failed read after processing already received data
+        // so that handshake failures are reported correctly
+        if (read == -1)
+            throw new EOFException("EOF during handshake");
```

Comment:
Hmm, so after processing already received data, isn't it possible that the handshake completed successfully? Maybe we should include the handshake status in the exception message.